### PR TITLE
feat: ability to associate timestamps with movements

### PR DESCRIPTION
**Pull request by aw1875**
Take two on a PR I did a few years ago (https://github.com/Xetera/ghost-cursor/pull/42).

This version does _essentially_ the same thing as the original PR but with a more elegant approach by utilizing the options that can be passed to the path function and calculating the time to take using the [Trapezoidal rule](https://en.wikipedia.org/wiki/Trapezoidal_rule).

I'm open to any suggestions/changes so please let me know.

### CODE REVIEW

**Comment by aw1875** [comment]
That's correct it's just for show. However, I wouldn't be against digging more into the `dispatchEvent` docs to use the generated timestamps in puppeteer if you think it would make sense for me to do.

**Comment by aw1875** [comment]
@Niek just checking back in so I don't lose track of this pr. Did you want me to make any changes?

**Comment by Niek** [comment]
In my opinion this only makes sense if the timestamps are used, meaning we need to switch to manual `Input.dispatchMouseEvent()` calls.

**Comment by aw1875** [comment]
> In my opinion this only makes sense if the timestamps are used, meaning we need to switch to manual `Input.dispatchMouseEvent()` calls.

Honestly, I don't really see any scenario where this would work unless I'm also making changes to functions like `move`, `moveTo`, etc. However, something like this _could technically_ take that approach (this is kinda hacky but you get the idea):

```diff
   const tracePath = async (
-    vectors: Iterable<Vector>,
+    vectors: Iterable<Vector | TimedVector>,
     abortOnMove: boolean = false
   ): Promise<void> => {
...
-        await page.mouse.move(v.x, v.y)
+
+        if ('timestamp' in v) {
+          await getCDPClient(page).send('Input.dispatchMouseEvent', {
+            type: 'mouseMoved',
+            x: v.x,
+            y: v.y,
+            button: 'none',
+            modifiers: 0,
+            timestamp: v.timestamp
+          })
+        } else {
+          await page.mouse.move(v.x, v.y)
+        }
+
...
```

Or, we could always use the dispatchMouseEvent with something more along the lines of:

```diff
   const tracePath = async (
-    vectors: Iterable<Vector>,
+    vectors: Iterable<Vector | TimedVector>,
     abortOnMove: boolean = false
   ): Promise<void> => {
...
-        await page.mouse.move(v.x, v.y)
+
+        const dispatchParams: Protocol.Input.DispatchMouseEventRequest = {
+          type: 'mouseMoved',
+          x: v.x,
+          y: v.y,
+          button: 'none',
+          modifiers: 0
+        }
+
+        if ('timestamp' in v) dispatchParams.timestamp = v.timestamp
+
+        await getCDPClient(page).send('Input.dispatchMouseEvent', dispatchParams)
...
```

**Comment by Niek** [comment]
Yes, this is what I was thinking of. Although it's probably better to have the CDP client initialised once in order to avoid creating thousands of clients.

**Comment by aw1875** [comment]
Good call, my brain didn't even process that I was calling that in a loop when I sent that code block

**Comment by Niek** [comment]
Awesome, thanks! This LGTM, but @bvandercar-vt can you review this as well?

**Comment by aw1875** [comment]
Not sure what happened there with the merge conflicts not getting all the changes the first time around and with the whitespace changes on the second one (husky ran properly)

**Comment by bvandercar-vt** [comment]
> Not sure what happened there with the merge conflicts not getting all the changes the first time around and with the whitespace changes on the second one (husky ran properly), if one of you could fix the whitespace I'd appreciate it.

Did you run `yarn lint`? Should fix the formatting

**Comment by aw1875** [comment]
> > Not sure what happened there with the merge conflicts not getting all the changes the first time around and with the whitespace changes on the second one (husky ran properly), if one of you could fix the whitespace I'd appreciate it.
> 
> Did you run `yarn lint`? Should fix the formatting

Yeah that was post `yarn lint` (plus husky should be running the linter automatically anyway)

File: src/spoof.ts [hunk 1]
```diff
@@ -76,6 +78,7 @@ export interface PathOptions {
    * Default is random.
    */
   readonly moveSpeed?: number
+  readonly showTimestamps?: boolean
```
[the remark this answers is not in the document]
Comment (aw1875) [reply src/spoof.ts:84]: Technically, we could have it always include a timestamp, but I feel like in most scenarios it's not necessary. I think just changing the name to something like `includeTimestamps` or even `useTimestamps` would make the most sense.

File: src/spoof.ts [hunk 2]
```diff
@@ -236,23 +239,65 @@ export function path (start: Vector, end: BoundingBox | Vector, optionsOrSpread?
...
+  return clampPositive(re, optionsResolved)
+}
+
+const clampPositive = (vectors: Vector[], options?: PathOptions): Vector[] | TimedVector[] => {
+  const clampedVectors = vectors.map((vector) => ({
+    x: Math.max(0, vector.x),
+    y: Math.max(0, vector.y)
+  }))
+
+  return options?.showTimestamps === false ? clampedVectors : generateTimestamps(clampedVectors, options)
```
Comment (bvandercar-vt) [inline src/spoof.ts:251]: Unless you intend timestamps to be included by default, this should be `return options?.showTimestamps ? generateTimestamps(clampedVectors, options) : clampedVectors`

**Comment by bvandercar-vt** [comment]
> > > Not sure what happened there with the merge conflicts not getting all the changes the first time around and with the whitespace changes on the second one (husky ran properly), if one of you could fix the whitespace I'd appreciate it.
> > 
> > 
> > Did you run `yarn lint`? Should fix the formatting
> 
> Yeah that was post `yarn lint` (plus husky should be running the linter automatically anyway)

Ah yeah `yarn lint` doesn't fix this whitespace issue on my end either:
image
Definitely a downside of `ts-standard` then. We could add `prettier` as a dep if the repo owners are interested.

File: src/spoof.ts (hunk 2 above)
Comment (aw1875) [reply src/spoof.ts:251]: Pretty sure I was explicit because the linter complains about it needing to be handled explicitly since it's optional.

Comment (bvandercar-vt) [reply src/spoof.ts:251]: That's super weird. Well then, either

```jsx
return options?.showTimestamps === true ? generateTimestamps(clampedVectors, options) : clampedVectors
```
or
```tsx
return options?.showTimestamps === false || options?.showTimestamps ===undefined ? clampedVectors : generateTimestamps(clampedVectors, options) 
```

Actually, does it work if you wrap in a `Boolean`?

```tsx
return Boolean(options?.showTimestamps) ? generateTimestamps(clampedVectors, options) : clampedVectors
```

Comment (bvandercar-vt) [reply src/spoof.ts:251]: The problem with the current `=== false` implementation is that if it's `undefined`, you're generating timestamps

**Comment by aw1875** [comment]
> > > > Not sure what happened there with the merge conflicts not getting all the changes the first time around and with the whitespace changes on the second one (husky ran properly), if one of you could fix the whitespace I'd appreciate it.
> > > 
> > > 
> > > Did you run `yarn lint`? Should fix the formatting
> > 
> > 
> > Yeah that was post `yarn lint` (plus husky should be running the linter automatically anyway)
> 
> Ah yeah `yarn lint` doesn't fix this whitespace issue on my end either:
> 
> image
> 
> Definitely a downside of `ts-standard` then. We could add `prettier` as a dep if the repo owners are interested.
> 
> Update: Not sure if the repo owners will go for it, but I wanted to see how it looked-- made a prettier PR: #142

super annoying, still a little what caused the whitespace to begin with as I turned off any formatters I have when working in this repo but not sure what else I can do besides manually fix it

**Comment by aw1875** [comment]
Should be good now, really odd but oh well

**Comment by bvandercar-vt** [comment]
@aw1875 PR looks good! Thanks for implementing those changes!

**Comment by Niek** [comment]
Thanks a lot, merged now!

**Pull request merged**